Check that a chromote connection is alive

hadley opened the pull request:
When your computer goes to sleep, the websocket and all debugging sessions are closed. This PR ensures that Chromote and ChromoteSession report this accurately.

I distinguished between active and alive for Chromote but not ChromoteSession since the browser process continues to run in the background, even though we have lost our connection to it. There's no way to restore a closed debugging session. However, I'm not sure about the names; maybe I have alive and active around the wrong way?

Part of #94

---

I tested this code with:

```R
library(chromote)

sess <- ChromoteSession$new()
sess$is_active()
#> [1] TRUE
sess$parent$is_alive()
#> [1] TRUE

# SLEEP

sess$is_active()
#> [1] FALSE
sess$parent$is_alive()
#> [1] FALSE
```

hadley commented:
Now Chromote will reconnect if the websocket is dead but the underlying process is alive:

```R
library(chromote)

chrome <- Chromote$new()
chrome$is_alive()
#> [1] TRUE
chrome$is_active()
#> [1] TRUE
chrome$check_active()

# SLEEP -----------------------

chrome$is_alive()
#> [1] TRUE
chrome$is_active()
#> [1] FALSE
chrome$check_active()
#> ! Reconnecting to chrome process.
#> ℹ All active sessions will be need to be respawned.
chrome$is_active()
#> [1] TRUE

chrome$close()
#> [1] FALSE
chrome$is_alive()
#> [1] FALSE
chrome$is_active()
#> [1] FALSE
chrome$check_active()
#> Error in `chrome$check_active()`:
#> ! Chromote has been closed.
```

wch commented on R/chromote.R:
```diff
@@ -335,13 +364,14 @@ Chromote <- R6Class(
     #' @description Close the [`Browser`] object
     close = function() {
-      if (private$is_active_) {
+      if (self$is_active()) {
+        private$ws$close()
+      }
+      if (self$is_alive()) {
         self$Browser$close()
```
I think this ordering of commands won't work properly. `self$Browser` is a domain of the Chrome Devtools Protocol, so `self$Browser$close()` sends the [`Browser.close`](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-close) command. But that won't work if the websocket has been closed.

hadley replied:
Oh yeah, I got confused. I thought this was `self$browser$close()`

wch commented:
Overall I'm OK with this change, but we will want to make it very clear to the user what the difference is between `is_alive` and `is_active`.

Perhaps the `print` method for `Chromote` could also print out that information? Like `Process is [not] alive` and `Websocket connection to Chrome is [not] active`.

And we'll also want to mark it as a breaking change in the NEWS file.

hadley commented:
I've addressed your concerns, apart from the print method, which I agree is a good idea and I'll do as part of #140, once this PR is merged.

Since your last review, I've added some additional logic to `ChromoteSession` to distinguish between the session and target being closed. This means that you get a reasonable error message if you `$close` the session (and hence the underlying target) and then try to `$respawn()` it. This is just a static value I cache in R, so it'll be incorrect if you close the target some other way, but I think that'll be uncommon and thus not worth handling. Let me know if you disagree.

```R
library(chromote)

sess <- ChromoteSession$new()
sess$is_active()
#> [1] TRUE

# SLEEP -----------------------

sess$is_active()
#> [1] FALSE

sess$check_active()
#> Error in `sess$check_active()`:
#> ! Session has been closed.
#> ℹ Call session$respawn() to create a new session that connects to the same target.

sess$close()
#> ! Reconnecting to chrome process.
#> ℹ All active sessions will be need to be respawned.

sess$respawn()
#> Error in `sess$respawn()`:
#> ! Can't respawn session; target has been closed.
#> [1] TRUE
```

wch commented on R/chromote_session.R:
```diff
@@ -535,19 +538,36 @@ ChromoteSession <- R6Class(
+    #'   as closed. For internal use only.
+    #' @param target_closed Has the underlying target been closed as well as the
+    #'   active debugging session?
+    mark_closed = function(target_closed) {
+      private$session_is_active <- FALSE
+      private$target_is_active <- FALSE
```
It looks like `target_closed` is unused?

hadley replied:
Doh. I realised that I forgot to call `spawn()` before closing in my script test. Here's the new output:

```R
library(chromote)

sess <- ChromoteSession$new()
sess$check_active()
#> NULL

# SLEEP -----------------------

sess$check_active()
#> Error in `sess$check_active()`:
#> ! Session has been closed.
#> ℹ Call session$respawn() to create a new session that connects to the same target.

sess2 <- sess$respawn()
#> ! Reconnecting to chrome process.
#> ℹ All active sessions will be need to be respawned.

sess$close()

sess$respawn()
#> Error in `sess$respawn()`:
#> ! Can't respawn session; target has been closed.
```

wch commented on R/chromote_session.R:
```diff
@@ -569,16 +589,15 @@ ChromoteSession <- R6Class(
     target_id = NULL,
-    is_active_ = NULL,
+    session_is_active = NULL,
+    target_is_active = NULL,
```
What do you think about `target_is_alive`? (I'm asking honestly, not suggesting that it should or should not be that name)

hadley replied:
I don't mind changing the name, but calling it `alive` does introduce another inconsistency — `Chromote$is_alive()` is always accurate because we have a connection to the process that we can check; but there's currently no analog for the target. I could add an `is_alive()` method that checks if the target ID is found in the chromote instance, but I'm not sure if that's worth it (plus calling that repeatedly might be expensive).

So while I agree that "active" isn't the best name, and has some potential for confusion, I think changing it to "alive" is a net neutral because it just introduces a different source confusion.

wch replied:
OK, let's keep it as "active" then.

wch commented:
I think the overall approach is good -- just have a couple questions.
